src/sentry/sentry_metrics/indexer/cloudspanner/cloudspanner.py

```diff
@@ -60,28 +118,315 @@ def validate(self) -> None:
+
+        def insert_uow(transaction: Any, rows: Sequence[SpannerIndexerModel]) -> None:
+            transaction.insert(
+                table=self._get_table_name(use_case_id), columns=get_column_names(), values=rows
+            )
+
+        try:
+            self.database.run_in_transaction(insert_uow, rows_to_insert)
+        except (
+            google.api_core.exceptions.AlreadyExists,
+            google.api_core.exceptions.InvalidArgument,
+        ):
+            raise CloudSpannerRowAlreadyExists
+        else:
+            metrics.incr(
+                _INDEXER_DB_INSERT_METRIC,
+                tags={"batch": "true"},
+            )
+            key_results.add_key_results(
+                [
+                    KeyResult(
+                        org_id=row.organization_id,
+                        string=row.string,
+                        id=row.decoded_id,
+                    )
+                    for row in rows_to_insert
+                ],
+                fetch_type=FetchType.FIRST_SEEN,
+            )
+
+    def _insert_collisions_handled(
+        self,
+        use_case_id: UseCaseKey,
+        rows_to_insert: Sequence[SpannerIndexerModel],
+        key_results: KeyResults,
+    ) -> None:
+        """
+        Insert the records in a transaction while handling collisions that
+        might happen during an INSERT. Collisions can happen  when multiple
+        consumers try to write records with same organization_id and string.
+
+        The basic logic is to retry the below in a loop until the transaction
+        succeeds.
+        The logic within a transaction is as follows:
+        1. Get records from the database for the organization_id and string
+        which are present in the rows.
+        2. Get records from the database for the id which are present in the rows.
+        3. Create a new list of records to be inserted which excludes the
+        records which are present in the above two lists.
+        4. Insert the new list of records in a transaction.
+
+        Once the transaction succeeds, we update the key_results by
+        performing a read of the organization_id and string which were
+        supposed to be originally inserted.
+        """
+
+        def insert_rw_transaction_uow(
+            transaction: Any, rows: Sequence[SpannerIndexerModel]
+        ) -> None:
+            existing_org_string_results = transaction.read(
+                table=self._get_table_name(use_case_id),
+                columns=["organization_id", "string", "id"],
+                keyset=spanner.KeySet(keys=[[row.organization_id, row.string] for row in rows]),
+                index=self._get_unique_org_string_index_name(use_case_id),
+            )
+
+            existing_id_results = transaction.read(
+                table=self._get_table_name(use_case_id),
+                columns=["organization_id", "string", "id"],
+                keyset=spanner.KeySet(keys=[[row.id] for row in rows]),
+            )
```

Comment:
I guess we cannot combine the two reads in one transactions because they are based on two separate indexes. IS that correct ?

Reply:
Yes, that's correct. It might be possible to run this in a singe transaction read with raw SQL but then it opens the door for SQL injection